File: src/Functional/Scala/Prototype.scala
final case class ServiceProfile(name: String, features: Vector[String]) {
  def cloneProfile: ServiceProfile = copy(features = features.map(identity))
  def describe: String = s"$name: ${features.mkString(",")}"
}

object Prototype extends App {
  val original = ServiceProfile("orders", Vector("metrics"))
  val baseClone = original.cloneProfile
  val canary = baseClone.copy(name = "orders-canary", features = baseClone.features :+ "tracing")

  println(s"original=${original.describe}")
  println(s"clone=${canary.describe}")
}

File: src/Functional/OCaml/prototype.ml
type service_profile = {
  name : string;
  features : string list;
}

let clone_profile profile =
  { profile with features = List.map Fun.id profile.features }

let describe profile =
  profile.name ^ ": " ^ String.concat "," profile.features

let () =
  let original = { name = "orders"; features = ["metrics"] } in
  let copy = clone_profile original in
  let canary =
    {
      name = "orders-canary";
      features = copy.features @ ["tracing"];
    }
  in
  Printf.printf "original=%s\n" (describe original);
  Printf.printf "clone=%s\n" (describe canary)
